lib/postgrex/extensions/tsvector.ex

```diff
@@ -0,0 +1,115 @@
+defmodule Postgrex.Extensions.Tsvector do
+  @moduledoc false
+
+  import Postgrex.BinaryUtils, warn: false
+  use Bitwise
+  use Postgrex.BinaryExtension, [send: "tsvectorsend"]
+
+  def encode(_) do
+    quote location: :keep do
+      values when is_list(values) ->
+        encoded_tsvector = unquote(__MODULE__).encode_tsvector(values)
+        <<byte_size(encoded_tsvector) :: int32, encoded_tsvector :: binary>>
+      other ->
+        raise ArgumentError,
+        Postgrex.Utils.encode_msg(other, "a list of tsvectors")
+    end
+  end
+
+  def decode(_) do
+    quote do
+      <<len :: int32, value :: binary-size(len)>> ->
+        <<nb_lexemes :: int32, words :: binary>> = value
+        unquote(__MODULE__).decode_tsvector_values(:binary.bin_to_list(words))
+    end
+  end
+
+  ## Helpers
+
+  def encode_tsvector(values) do
+    << length(values) :: int32, encode_lexemes(values) :: binary >>
+  end
+
+  def encode_lexemes(values) do
+    Enum.map(values, fn(%Postgrex.Lexeme{word: word, positions: positions}) ->
+      encode_positions("#{word}\0", positions)
+    end)
+    |> IO.iodata_to_binary
+  end
+
+  defp encode_positions(bytes, position) do
+      positions = Enum.map(position, fn(x) ->
+        chars = to_char_list(x)
+        << encode_weight(chars)::2, encode_position(chars)::14 >>
+      end)
+                  |> IO.iodata_to_binary
+      [bytes, <<length(position) :: 16>>, positions]
+  end
+
+  defp encode_weight(data) do
+    Enum.find(data, -1, fn(x) -> (x >= 65 && x <= 67) || (x >= 97 && x <= 99) end)
+    |> encode_weight_binary
+  end
+
+  defp encode_position(data) do
+   Enum.filter(data, fn(x) -> x >= 48 && x <= 57 end)
+   |> :string.to_integer
+   |> elem(0)
```

Comment:
I've converted the whole encode part to {position, weight} with charlist support.

It allows us to trim all the Enum.filter part in this code.